### PR TITLE
using hash_file for lists and tuples (closes #178)

### LINKED ISSUE #178: list of Files as an input
`BaseSpec.hash` will not work properly for list of `File`s.

### PR DESCRIPTION
**Pull request by djarecka**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Summary

calling `hash_file` if type is ty.List[File] or ty.Tuple[File], `hash_file` can handle lists/tuples

## Checklist

- [ ] All tests passing
- [ ] I have added tests to cover my changes
- [ ] I have updated documentation (if necessary)
- [x] My code follows the code style of this project 
(we are using `black`: you can `pip install pre-commit`, 
run `pre-commit install` in the `pydra` directory 
and `black` will be run automatically with each commit)

## Acknowledgment
- [x] I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

File: pydra/engine/helpers_file.py [hunk 1]
```diff
@@ -113,7 +113,7 @@ def hash_file(afile, chunk_len=8192, crypto=sha256, raise_notfound=True):
     from .helpers import hash_function
 
     # adding option for tasks with splitter over list of files
-    if isinstance(afile, list):
+    if isinstance(afile, (list, tuple)):
```
Comment (satra) [inline pydra/engine/helpers_file.py]: how do you know this list or tuple contains files?

Comment (satra) [reply pydra/engine/helpers_file.py]: nevermind

File: pydra/engine/specs.py [hunk 2]
```diff
@@ -50,6 +50,8 @@ def hash(self):
                 continue
             if field.type == File and "container_path" not in field.metadata:
                 inp_dict[field.name] = hash_file(getattr(self, field.name))
+            elif field.type == ty.List[File] or field.type == ty.Tuple[File]:
```
Comment (satra) [inline pydra/engine/specs.py]: what if it was a `List[List[File]`?

it's likely that this bit has to be refactored into a recursive function call. also should support dictionaries whose values are `File`.  this is the function from nipype:

https://github.com/nipy/nipype/blob/master/nipype/interfaces/base/specs.py#L273

i don't think some of the elements are necessary with py3, but the general idea is still something to consider here.

**Comment by djarecka** [comment]
created #195 instead

**Pull request closed without merging**